include/ck/tensor_operation/gpu/grid/gridwise_batched_mha_bwd_xdl_cshuffle_qloop_b2t_v2.hpp

```diff
@@ -467,16 +489,123 @@ struct GridwiseBatchedMultiheadAttentionBackward_Qloop_Xdl_CShuffle_V2
+        using KBlockwiseCopy =
+            ThreadGroupTensorSliceTransfer_v4r1<ThisThreadBlock,
+                                                tensor_operation::element_wise::PassThrough,
+                                                tensor_operation::element_wise::PassThrough,
+                                                InMemoryDataOperationEnum::Set,
+                                                Sequence<K_K0, NPerBlock, BK1>,
+                                                BBlockTransferThreadClusterLengths_BK0_N_BK1,
+                                                BBlockTransferThreadClusterArrangeOrder,
+                                                InputDataType,
+                                                GemmDataType,
+                                                GridDesc_K0_N_K1,
+                                                decltype(k_block_desc_k0_n_k1),
+                                                BBlockTransferSrcAccessOrder,
+                                                Sequence<1, 0, 2>,
+                                                BBlockTransferSrcVectorDim,
+                                                2,
+                                                BBlockTransferSrcScalarPerVector,
+                                                BBlockTransferDstScalarPerVector_BK1,
+                                                1,
+                                                1,
+                                                true, // SrcResetCoord
+                                                true, // DstResetCoord
+                                                NumGemmKPrefetchStage>;
+
+        template <typename GridDesc_K0_K1_k2_N0_N1_N2_N3_K3>
+        using VBlockwiseCopy = ThreadwiseTensorSliceTransfer_v2<
+            InputDataType,
+            GemmDataType,
+            GridDesc_K0_K1_k2_N0_N1_N2_N3_K3,
+            decltype(v_thread_desc_k0_k1_k2_n0_n1_n2_n3_k3),
+            decltype(v_thread_desc_k0_k1_k2_n0_n1_n2_n3_k3.GetLengths()),
+            Sequence<0, 1, 2, 3, 4, 5, 6, 7>,
+            7,
+            BK1,
+            1,
+            true /* ResetCoordAfterRun */>;
+
+        static constexpr auto VBlockBufferSize        = V_K0;
+        static constexpr auto v_block_slice_copy_step = make_multi_index(1, 0, 0, 0, 0, 0, 0, 0);
+    };
+
+    // dP Gemm (type 1 rcc, B in Vgpr)
     struct Gemm0
     {
         // A matrix in LDS memory, dst of blockwise copy
         static constexpr auto a_block_desc_ak0_m_ak1 =
             GetABlockDescriptor_AK0PerBlock_MPerBlock_AK1();
 
-        // B matrix in LDS memory, dst of blockwise copy
-        static constexpr auto b_block_desc_bk0_n_bk1 =
-            GetBBlockDescriptor_BK0PerBlock_NPerBlock_BK1();
+        // B source matrix layout in VGPR
+        static constexpr auto b_src_thread_desc_k0_k1_k2_n0_n1_n2_n3_k3 =
+            GetVThreadDescriptor_K0_K1_K2_N0_N1_N2_N3_K3();
```

Comment:
It will be better if you put `BSrcThreadDesc_K0_K1_K3_N0_N1_N2_N3_K3` as template of `Gemm0`,  so that you don't have to put `GetVThreadDescriptor_xx` inside the codes of `struct Gemm0`,   refer to `struct Gemm1`